File: src/de/saschahlusiak/freebloks/view/BackgroundRenderer.java
package de.saschahlusiak.freebloks.view;

import javax.microedition.khronos.opengles.GL10;
import javax.microedition.khronos.opengles.GL11;

import de.saschahlusiak.freebloks.view.model.Theme;


import android.content.res.Resources;
import android.graphics.Bitmap;
import android.opengl.GLUtils;

public class BackgroundRenderer extends SimpleModel {
	float rgba[] = { 0, 0, 0, 1 };
	int texture[];
	boolean hasTexture;

    final static int num_vertices = 4;
    final static int num_triangles = 2;
    final static float size = 80.0f;
    final static float textures = 15.0f;

    Theme theme;
    Resources resources;

    public BackgroundRenderer(Resources resources) {
		super(num_vertices, num_triangles);

		this.resources = resources;
		hasTexture = false;

	    addVertex(-size, 0, -size, 0, 1, 0, 0, 0);
	    addVertex( size, 0, -size, 0, 1, 0, textures, 0);
	    addVertex( size, 0,  size, 0, 1, 0, textures, textures);
	    addVertex(-size, 0,  size, 0, 1, 0, 0, textures);

	    addIndex(0, 2, 1);
	    addIndex(0, 3, 2);

	    commit();
	}

	boolean valid = false;

	public void setTheme(Theme theme) {
		this.theme = theme;
		valid = false;
	}

	public void updateTexture(GL10 gl) {
		valid = true;
		if (theme == null) {
			rgba[0] = 0.0f; // transparent
			rgba[1] = 0.0f;
			rgba[2] = 0.0f;
			rgba[3] = 0.0f;
			hasTexture = false;
		} else if (theme.isDrawable()) {
			hasTexture = true;

			texture = new int[1];
			gl.glGenTextures(1, texture, 0);

			gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[0]);
		  	if (gl instanceof GL11) {
				gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR_MIPMAP_LINEAR);
		  		gl.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_GENERATE_MIPMAP, GL11.GL_TRUE);
		  	} else {
				gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR);
		  		gl.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_GENERATE_MIPMAP, GL11.GL_FALSE);
		  	}
		  	gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR);
		  	FreebloksRenderer.loadKTXTexture(gl, resources, theme.getTexture());

			rgba = theme.getRGBA();
		} else {
			hasTexture = false;
			rgba = theme.getRGBA();
		}
	}

	final float specular[] = {0, 0, 0, 1};

	public void render(GL11 gl) {
		if (!valid)
			updateTexture(gl);

		if (hasTexture)
			gl.glClearColor(0.05f, 0.10f, 0.25f, 1.0f); /* the default background when textured */
		else
			gl.glClearColor(rgba[0], rgba[1], rgba[2], rgba[3]);

		gl.glClear(GL10.GL_COLOR_BUFFER_BIT | GL10.GL_DEPTH_BUFFER_BIT);

		if (hasTexture) {
//			float shininess[]={3.0f};

			gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, rgba, 0);
			gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, specular, 0);
//			gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, shininess, 0);

			gl.glEnable(GL10.GL_TEXTURE_2D);

			gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[0]);
			bindBuffers(gl);

			drawElements(gl);

			gl.glDisable(GL10.GL_TEXTURE_2D);
		}
	}
}

File: app/src/main/java/de/saschahlusiak/freebloks/view/BackgroundRenderer.kt
package de.saschahlusiak.freebloks.view

import android.content.res.Resources
import de.saschahlusiak.freebloks.ktx.KTX
import de.saschahlusiak.freebloks.theme.Theme
import javax.microedition.khronos.opengles.GL10
import javax.microedition.khronos.opengles.GL11

class BackgroundRenderer(private val resources: Resources, private var theme: Theme) : SimpleModel(4, 2, false) {
    private val rgba = floatArrayOf(0f, 0f, 0f, 1f)
    private val specular = floatArrayOf(0f, 0f, 0f, 1f)
    private val size = 80.0f
    private val textures = 15.0f

    private var texture: IntArray? = null
    private var hasTexture = false
    private var valid = false

    init {
        addVertex(-size, 0f, -size, 0f, 1f, 0f, 0f, 0f)
        addVertex(size, 0f, -size, 0f, 1f, 0f, textures, 0f)
        addVertex(size, 0f, size, 0f, 1f, 0f, textures, textures)
        addVertex(-size, 0f, size, 0f, 1f, 0f, 0f, textures)
        addIndex(0, 2, 1)
        addIndex(0, 3, 2)

        commit()
    }

    fun setTheme(theme: Theme) {
        this.theme = theme
        valid = false
    }

    fun updateTexture(gl: GL11) {
        valid = true
        val asset = theme.asset

        if (theme.isResource && asset != null) {
            hasTexture = true
            texture = IntArray(1).also { texture ->
                gl.glGenTextures(1, texture, 0)
                gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[0])
            }

            gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR_MIPMAP_LINEAR)
            gl.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_GENERATE_MIPMAP, GL11.GL_TRUE.toFloat())
            gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR)

            KTX.loadKTXTexture(resources.assets, asset)

            theme.getColor(resources, rgba)
            rgba[3] = 1.0f
        } else {
            hasTexture = false
            texture = null
            theme.getColor(resources, rgba)
            rgba[3] = 1.0f
        }
    }

    fun render(gl: GL11) {
        if (!valid) updateTexture(gl)

        if (hasTexture)
            gl.glClearColor(0.0f, 0.0f, 0.0f, 1.0f)
        else
            gl.glClearColor(rgba[0], rgba[1], rgba[2], rgba[3])

        gl.glClear(GL10.GL_COLOR_BUFFER_BIT or GL10.GL_DEPTH_BUFFER_BIT)

        if (hasTexture) {
            texture?.let { texture ->
                with(gl) {
                    glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, rgba, 0)
                    glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, specular, 0)
                    glEnable(GL10.GL_TEXTURE_2D)
                    glBindTexture(GL10.GL_TEXTURE_2D, texture[0])
                    bindBuffers(gl)
                    drawElements(gl, GL10.GL_TRIANGLES)
                    glDisable(GL10.GL_TEXTURE_2D)
                }
            }
        }
    }
}